src/script/sign.h

```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+        }
+
+        // Separator
+        s << PSBT_SEPARATOR;
+
+        // Write inputs
+        for (const PartiallySignedInput& input : inputs) {
+            s << input;
+        }
+        // Write outputs
+        for (const PSBTOutput& output : outputs) {
+            s << output;
+        }
+    }
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        // Read the magic bytes
+        int magic;
+        s >> magic;
+        if (magic != PSBT_MAGIC_BYTES) {
+            throw std::ios_base::failure("Invalid PSBT magic bytes");
+        }
+        unsigned char magic_sep;
+        s >> magic_sep;
+
+        // Read global data
+        while(!s.empty()) {
+            // read size of key
+            uint64_t key_len = ReadCompactSize(s);
+
+            // the key length is 0 if that was actually a separator byte
+            // This is a special case for key lengths 0 as those are not allowed (except for separator)
+            if (key_len == 0) {
+                break;
+            }
+
+            // Read key
+            std::vector<unsigned char> key;
+            key.resize(key_len);
+            s >> MakeSpan(key);
+
+            // First byte of key is the type
+            unsigned char type = key[0];
+
+            // Do stuff based on type
+            switch(type) {
+                case PSBT_GLOBAL_UNSIGNED_TX:
+                    if (!CTransaction(tx).IsNull()) {
+                        throw std::ios_base::failure("Duplicate Key, unsigned tx already provided");
+                    }
+                    UnserializeFromVector(s, tx);
+                    // Make sure that all scriptSigs and scriptWitnesses are empty
+                    for (const CTxIn& txin : tx.vin) {
+                        if (!txin.scriptSig.empty() || !txin.scriptWitness.IsNull()) {
+                            throw std::ios_base::failure("Unsigned tx does not have empty scriptSigs and scriptWitnesses.");
+                        }
+                    }
+                    break;
+                // Unknown stuff
+                default: {
+                    if (unknown.count(key) > 0) {
+                        throw std::ios_base::failure("Duplicate Key, key for unknown value already provided");
+                    }
+                    // Read in the value
+                    uint64_t value_len = ReadCompactSize(s);
+                    std::vector<unsigned char> val_bytes;
+                    val_bytes.resize(value_len);
+                    s >> MakeSpan(val_bytes);
+                    unknown.emplace(std::move(key), std::move(val_bytes));
+                }
+            }
+        }
+
+        // Make sure that we got an unsigned tx
+        if (CTransaction(tx).IsNull()) {
+            throw std::ios_base::failure("No unsigned transcation was provided");
+        }
+
+        // Read input data
+        unsigned int i = 0;
+        while (!s.empty() && i < tx.vin.size()) {
```

Comment:
Why the `!s.empty()` check here? If we reach EOF in the stream an error should be raised, not ignore it.

Reply:
If we reach EOF in the stream, this loop will exit and the if block below will throw the error as the number of inputs/outputs will differ from the number of inputs/outputs in the unsigned tx.

Reply:
Ah, I see. That will probably give a more useful error.